File: dummy_udp_vehicle.m
function dummy_udp_vehicle(managerIP, odometryPort, commandPort, odometryType, rateHz)
%DUMMY_UDP_VEHICLE Standalone loopback transport test with a 0.25 s command timeout.
% dummy_udp_vehicle("127.0.0.1",12345,34567,"speed&angularVelocity",20)
    if nargin<1, managerIP="127.0.0.1"; end
    if nargin<2, odometryPort=12345; end
    if nargin<3, commandPort=34567; end
    if nargin<4, odometryType="speed&angularVelocity"; end
    if nargin<5, rateHz=20; end
    validateattributes(rateHz,{'numeric'},{'scalar','positive','finite'});
    assert(ismember(string(odometryType),["position&orientation","speed&angularVelocity"]), ...
        'Station:InvalidSettings','Unsupported UDP odometry type.');
    odometryTx=udpport("datagram","IPV4","ByteOrder","big-endian");
    commandRx=udpport("datagram","IPV4","LocalPort",commandPort,"ByteOrder","big-endian");
    cleaner=onCleanup(@() release(odometryTx,commandRx));
    fprintf('Dummy UDP: state -> %s:%d, command port=%d. Ctrl+C to stop.\n', ...
        managerIP,odometryPort,commandPort);
    x=0; y=0; theta=0; v=0; w=0; seq=0;
    received=[];
    previous=tic;
    while true
        step=tic;
        dt=toc(previous);
        previous=tic;
        count=commandRx.NumDatagramsAvailable;
        if count>0
            packets=read(commandRx,count,"uint8");
            for k=1:numel(packets)
                bytes=uint8(packets(k).Data);
                if numel(bytes)~=16, continue; end
                command=typecast(bytes(:),'double');
                [~,~,endian]=computer;
                if endian=='L', command=swapbytes(command); end
                if all(isfinite(command))
                    v=command(1); w=command(2); received=tic;
                end
            end
        end
        if isempty(received) || toc(received)>0.25
            v=0; w=0;
        end
        x=x+v*cos(theta)*dt;
        y=y+v*sin(theta)*dt;
        theta=theta+w*dt;
        seq=seq+1;
        if odometryType=="position&orientation"
            packet=[x y theta];
        else
            packet=[seq v w];
        end
        write(odometryTx,packet,"double",managerIP,odometryPort);
        pause(max(0,1/rateHz-toc(step)));
    end
end

function release(tx,rx)
    delete(rx);
    delete(tx);
end

File: dummy_udp_vehicle.py
#!/usr/bin/env python3
"""Simple UDP dummy vehicle for manager-side integration tests.

Example:
    python dummy_udp_vehicle.py --manager-ip 127.0.0.1 --odometry-port 12345 \
        --command-port 23456 --odometry-type "speed&angularVelocity" --rate-hz 20
"""

from __future__ import annotations

import argparse
import math
import socket
import struct
import time


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="UDP dummy vehicle")
    parser.add_argument("--manager-ip", default="127.0.0.1")
    parser.add_argument("--odometry-port", type=int, default=12345)
    parser.add_argument("--command-port", type=int, default=34567)
    parser.add_argument(
        "--odometry-type",
        default="speed&angularVelocity",
        choices=["position&orientation", "speed&angularVelocity", "velocity&angularVelocity"],
    )
    parser.add_argument("--rate-hz", type=float, default=20.0)
    return parser.parse_args()


def pack_doubles(values: list[float]) -> bytes:
    return struct.pack(">" + "d" * len(values), *values)  # big-endian doubles


def unpack_doubles(data: bytes) -> list[float]:
    count = len(data) // 8
    if count == 0:
        return []
    return list(struct.unpack(">" + "d" * count, data[: count * 8]))


def main() -> None:
    args = parse_args()

    odom_tx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    cmd_rx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    cmd_rx.bind(("", args.command_port))
    cmd_rx.setblocking(False)

    print(
        "Dummy UDP vehicle started: "
        f"odometry->{args.manager_ip}:{args.odometry_port}, "
        f"commandPort={args.command_port}, type={args.odometry_type}"
    )
    print("Press Ctrl+C to stop.")

    dt = 1.0 / args.rate_hz
    x = 0.0
    y = 0.0
    theta = 0.0
    v = 0.5
    w = 0.3
    seq = 0

    try:
        while True:
            start = time.perf_counter()
            seq += 1

            if args.odometry_type == "position&orientation":
                x += v * math.cos(theta) * dt
                y += v * math.sin(theta) * dt
                theta += w * dt
                packet = [x, y, theta]
            else:
                # Keep MATLAB compatibility for speed mode:
                # [seq, speed, angularVelocity]
                packet = [float(seq), v, w]

            odom_tx.sendto(pack_doubles(packet), (args.manager_ip, args.odometry_port))

            # Optional command read for visibility during testing.
            while True:
                try:
                    data, _ = cmd_rx.recvfrom(4096)
                except BlockingIOError:
                    break
                command = unpack_doubles(data)
                if command:
                    print(f"Received command: {command}")

            elapsed = time.perf_counter() - start
            sleep_time = dt - elapsed
            if sleep_time > 0:
                time.sleep(sleep_time)
    except KeyboardInterrupt:
        print("Stopping dummy UDP vehicle.")
    finally:
        cmd_rx.close()
        odom_tx.close()


if __name__ == "__main__":
    main()
